File: problems/0004_median-of-two-sorted-arrays/binary-search/solution.c
#define MIN(X,Y) ((X) < (Y) ? (X) : (Y))
#define MAX(X,Y) ((X) > (Y) ? (X) : (Y))
#define ERROR(fmt, ...) fprintf(stderr, fmt, ##__VA_ARGS__)

double findMedianSortedArrays(int* nums1, int nums1Size, int* nums2, int nums2Size) {
    int *A = nums1, *B = nums2, m = nums1Size, n = nums2Size;
    if (nums1Size > nums2Size) {
        A = nums2;
        m = nums2Size;
        B = nums1;
        n = nums1Size;
    }
    const int total = m + n,
              h = total / 2;
    const bool isOdd = (total % 2 == 1);
    for (int iBegin = 0, iEnd = m; iBegin <= iEnd;) {
        const int i = (iBegin + iEnd) / 2,
                  j = h - i;
        if (i < iEnd && 0 < j && A[i] < B[j - 1])
            iBegin = i + 1;
        else if (iBegin < i && j < n && B[j] < A[i - 1])
            iEnd = i - 1;
        else {
            const int minGreater = (i == m ?
                    B[j] :
                    (j == n ?
                        A[i] :
                        MIN(A[i], B[j])));
            if (isOdd)
                return minGreater;
            const int maxLesser = (i == 0 ?
                    B[j - 1] :
                    (j == 0 ?
                        A[i - 1] :
                        MAX(A[i - 1], B[j - 1])));
            return (maxLesser + minGreater) / 2.0;
        }
    }
    ERROR("Should NOT be HERE!");
    return NAN;
}

File: problems/0004_median-of-two-sorted-arrays/binary-search/solution.ts
function findMedianSortedArrays(nums1: number[], nums2: number[]): number {
    const [A, B] = (nums1.length <= nums2.length ? [nums1, nums2] : [nums2, nums1]),
        m = A.length,
        n = B.length,
        total = m + n,
        h = total >> 1,
        isOdd = (total % 2 === 1)
    for (let iBegin = 0, iEnd = m; iBegin <= iEnd;) {
        const i = (iBegin + iEnd) >> 1,
            j = h - i
        if (i < iEnd && 0 < j && A[i] < B[j - 1]) // A[i] is too small
            iBegin = i + 1
        else if (iBegin < i && j < n && B[j] < A[i - 1]) // A[i-1] is too large
            iEnd = i - 1
        else {
            const minGreater = (i === m ? // A is empty OR all A[i] < B[j-1]
                    B[j] :
                    (j === n ? // m == n AND all B[j] < A[i-1]
                        A[i] :
                        Math.min(A[i], B[j])))
            if (isOdd)
                return minGreater
            const maxLesser = (i === 0 ? // A is empty OR all B[j] < A[i-1]
                    B[j - 1] :
                    (j === 0 ? // h == i AND all A[i] < B[j-1]
                        A[i - 1] :
                        Math.max(A[i - 1], B[j - 1])))
            return (maxLesser + minGreater) / 2
        }
    }
    throw `Should NOT be HERE!`
};
